gcc/rust/expand/rust-expand-visitor.cc

```diff
@@ -18,38 +18,199 @@
+ * &attr2}]
+ * ```
+ *
+ * @param outer_attrs The list of attributes on the item to derive
+ */
+static std::vector<
+  std::pair<std::string, std::reference_wrapper<const AST::Attribute>>>
+get_traits_to_derive (std::vector<AST::Attribute> &outer_attrs)
+{
+  std::vector<
+    std::pair<std::string, std::reference_wrapper<const AST::Attribute>>>
+    to_derive;
+  for (auto it = outer_attrs.begin (); it != outer_attrs.end ();)
+    {
+      auto &attr = *it;
+
+      if (is_derive (attr))
+	{
+	  auto &input = attr.get_attr_input ();
+	  switch (input.get_attr_input_type ())
+	    {
+	      // isn't there a better way to do this?? like parse it or
+	      // something idk. some function I'm not thinking of?
+	      case AST::AttrInput::TOKEN_TREE: {
+		auto &tokens = static_cast<AST::DelimTokenTree &> (input)
+				 .get_token_trees ();
+
+		// erase the delimiters
+		rust_assert (tokens.size () >= 3);
+		tokens.erase (tokens.begin ());
+		tokens.pop_back ();
+
+		for (auto &token : tokens)
+		  {
+		    // skip commas, as they are part of the token stream
+		    if (token->as_string () == ",")
+		      continue;
+
+		    to_derive.emplace_back (token->as_string (), attr);
+		  }
+		break;
+	      }
+	    case AST::AttrInput::LITERAL:
+	    case AST::AttrInput::META_ITEM:
+	    case AST::AttrInput::MACRO:
+	      gcc_unreachable ();
+	      break;
+	    }
+
+	  it = outer_attrs.erase (it);
+	}
+      else
+	{
+	  it++;
+	}
+    }
+
+  return to_derive;
+}
+
+static std::unique_ptr<AST::Item>
+derive_item (std::unique_ptr<AST::Item> &item, const AST::Attribute &derive,
+	     BuiltinMacro to_derive)
+{
+  return AST::DeriveVisitor::derive (*item, derive, to_derive);
+}
+
+void
+ExpandVisitor::expand_inner_items (
+  std::vector<std::unique_ptr<AST::Item>> &items)
 {
   expander.push_context (MacroExpander::ContextType::ITEM);
 
-  // expand attributes recursively and strip items if required
-  //  AttrVisitor attr_visitor (*this);
-  auto &items = crate.items;
-  for (auto it = items.begin (); it != items.end ();)
+  for (auto it = items.begin (); it != items.end (); it++)
     {
       auto &item = *it;
-      item->accept_vis (*this);
-
-      auto fragment = expander.take_expanded_fragment ();
-      if (fragment.should_expand ())
+      if (item->has_outer_attrs ())
 	{
-	  // Remove the current expanded invocation
-	  it = items.erase (it);
-	  for (auto &node : fragment.get_nodes ())
+	  auto traits_to_derive
+	    = get_traits_to_derive (item->get_outer_attrs ());
+
+	  for (auto &to_derive : traits_to_derive)
 	    {
-	      it = items.insert (it, node.take_item ());
-	      it++;
+	      auto &name = to_derive.first;
+	      auto &attr = to_derive.second;
+
+	      auto maybe_builtin = MacroBuiltin::builtins.lookup (name);
+	      if (MacroBuiltin::builtins.is_iter_ok (maybe_builtin))
+		{
+		  auto new_item
+		    = derive_item (item, attr, maybe_builtin->second);
+		  // this inserts the derive *before* the item - is it a
```

Comment:
rustc inserts the functions after the item

Reply:
shouldn't be an issue, but we'll see in the future if it arises. it's not too hard of a fix either way :)